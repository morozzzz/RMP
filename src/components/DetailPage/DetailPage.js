import React from 'react';
import MovieList from '../MovieList/MovieList';
import StatusBar from '../StatusBar/StatusBar';
import DetailMovieBlock from '../DetailMovieBlock/DetailMovieBlock';
import Footer from '../Footer/Footer';
import ErrorBoundary from '../ErrorBoundary/ErrorBoundary';
import MovieService from '../../services/MovieService';
import '@babel/polyfill';
import './DetailPage.css';

const getStatus = (text) => {
    if (!text) {
        return '';
    }

    return `Films by ${text.charAt(0).toUpperCase() + text.slice(1)} genre`;
};

class DetailPage extends React.Component {
    constructor({ detailedMovie }) {
        super();
        this.state = {
            detailedMovie: detailedMovie || {
                id: '',
                title: '',
                tagline: '',
                vote_average: '',
                vote_count: '',
                release_date: '',
                poster_path: '',
                overview: '',
                budget: '',
                revenue: '',
                genres: [''],
                runtime: '',
            },
            movies: [],
        };
    }

    onPosterClick = async (data) => {
        const movies = await MovieService.getMovies(data.genres[0]);
        this.setState({ movies, detailedMovie: data });
    }

    componentDidMount = () => {
        const { detailedMovie } = this.state;
        const genre = detailedMovie && detailedMovie.genres[0];

        MovieService.getMovies({ genre })
            .then((movies) => {
                this.setState({ movies });
            });
    }

    render() {
        const { movies, detailedMovie } = this.state;
        const singleGenre = detailedMovie.genres[0];

        return (
            <ErrorBoundary>
                <div className="detail-page">
                    <DetailMovieBlock movieData={detailedMovie} />
                    <StatusBar status={getStatus(singleGenre)} />
                    <MovieList movies={movies} onItemClick={this.onPosterClick} />
                    <Footer />
                </div>
            </ErrorBoundary>
        );
    }
}

export default DetailPage;
