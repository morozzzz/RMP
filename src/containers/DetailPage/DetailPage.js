import React from 'react';
import { connect } from 'react-redux';
import MovieList from '../../components/MovieList/MovieList';
import StatusBar from '../../components/StatusBar/StatusBar';
import DetailMovieBlock from '../../components/DetailMovieBlock/DetailMovieBlock';
import Footer from '../../components/Footer/Footer';
import ErrorBoundary from '../../components/ErrorBoundary/ErrorBoundary';
import { fetchDetailedMovie } from '../../actions/movies.actions';
import '@babel/polyfill';
import './DetailPage.css';

const getStatus = (text) => {
    if (!text) {
        return '';
    }

    return `Films by ${text.charAt(0).toUpperCase() + text.slice(1)} genre`;
};

class DetailPage extends React.Component {
    onPosterClick = async (data) => {
        const { getDetailedMovie, history } = this.props;

        history.push(`/film/${data.id}`);

        getDetailedMovie(data.id);
    }

    componentDidMount = () => {
        const { getDetailedMovie, match } = this.props;
        const { params } = match;
        const movieId = params.id;

        getDetailedMovie(movieId);
    }

    render() {
        const { movies, detailedMovie } = this.props;
        const singleGenre = detailedMovie && detailedMovie.genres && detailedMovie.genres[0];

        return (
            <ErrorBoundary>
                <div className="detail-page">
                    <DetailMovieBlock movieData={detailedMovie} />
                    <StatusBar status={getStatus(singleGenre)} />
                    {
                        detailedMovie
                            ? <MovieList movies={movies} onItemClick={this.onPosterClick} />
                            : <div />
                    }
                    <Footer />
                </div>
            </ErrorBoundary>
        );
    }
}

const mapStateToProps = state => (
    {
        detailedMovie: state.movies.detailed,
        movies: state.movies.related,
    }
);

const mapDispatchToProps = dispatch => (
    {
        getDetailedMovie: id => dispatch(fetchDetailedMovie(id)),
    }
);

export default connect(mapStateToProps, mapDispatchToProps)(DetailPage);
