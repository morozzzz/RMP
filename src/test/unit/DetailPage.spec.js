import React from 'react';
import { shallow } from 'enzyme';
import DetailPage from '../../containers/DetailPage/DetailPage';

jest.mock('../../services/MovieService');

const detailedMovieMock = {
    id: 338970,
    title: 'Tomb Raider',
    tagline: 'Her legend begins',
    vote_average: 6.2,
    vote_count: 817,
    release_date: '2018-03-08',
    poster_path: 'https://image.tmdb.org/t/p/w500/ePyN2nX9t8SOl70eRW47Q29zUFO.jpg',
    overview: 'Lara Croft, the fiercely independent daughter of a missing adventurer, must push herself beyond her limits when she finds herself on the island where her father disappeared.',
    budget: 94000000,
    revenue: 126025000,
    genres: [
        'Action',
        'Adventure',
    ],
    runtime: 118,
};

jest.mock('../../components/MovieList/MovieList', () => () => <div id="MovieList" />);
jest.mock('../../components/StatusBar/StatusBar', () => () => <div id="StatusBar" />);
jest.mock('../../components/DetailMovieBlock/DetailMovieBlock', () => () => <div id="DetailMovieBlock" />);
jest.mock('../../components/Footer/Footer', () => () => <div id="Footer" />);

describe('DetailPage', () => {
    it('component should be rendered correctly', () => {
        const component = shallow(<DetailPage detailedMovie={detailedMovieMock} />);

        expect(component).toMatchSnapshot();
    });
});
