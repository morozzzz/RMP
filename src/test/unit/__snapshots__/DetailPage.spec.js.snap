// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailPage component should be rendered correctly 1`] = `
<ErrorBoundary>
  <div
    className="detail-page"
  >
    <Component
      movieData={
        Object {
          "budget": 94000000,
          "genres": Array [
            "Action",
            "Adventure",
          ],
          "id": 338970,
          "overview": "Lara Croft, the fiercely independent daughter of a missing adventurer, must push herself beyond her limits when she finds herself on the island where her father disappeared.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ePyN2nX9t8SOl70eRW47Q29zUFO.jpg",
          "release_date": "2018-03-08",
          "revenue": 126025000,
          "runtime": 118,
          "tagline": "Her legend begins",
          "title": "Tomb Raider",
          "vote_average": 6.2,
          "vote_count": 817,
        }
      }
    />
    <Component
      status="Films by Action genre"
    />
    <Component
      movies={Array []}
      onItemClick={[Function]}
    />
    <Component />
  </div>
</ErrorBoundary>
`;
